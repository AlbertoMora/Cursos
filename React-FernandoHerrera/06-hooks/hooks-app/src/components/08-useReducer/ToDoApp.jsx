import React, { useEffect, useReducer } from "react";

import "./styles.css";
import { todoReducer } from "./todoReducer";
import ToDoList from "./ToDoList";
import ToDoAdd from "./ToDoAdd";

const init = () => {
    return JSON.parse(localStorage.getItem("todos")) || [];
};

const ToDoApp = () => {
    const [toDos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(toDos));
    }, [toDos]);

    const handleDelete = (todoId) => {
        const deleteToDoAction = {
            type: "remove",
            payload: todoId,
        };
        dispatch(deleteToDoAction);
    };

    const handleToggle = (todoId) => {
        const toggleToDoAction = {
            type: "toggle",
            payload: todoId,
        };
        dispatch(toggleToDoAction);
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: "add",
            payload: newTodo,
        });
    }

    return (
        <>
            <h1>Todo App ({toDos.filter((todo) => todo.done === false).length})</h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <ToDoList toDos={toDos} handleDelete={handleDelete} handleToggle={handleToggle} />
                </div>
                <div className="col-5">
                   <ToDoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </>
    );
};

export default ToDoApp;
