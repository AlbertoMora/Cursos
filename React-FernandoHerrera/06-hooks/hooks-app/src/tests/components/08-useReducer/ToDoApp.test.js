import React from "react";
import { shallow, mount } from "enzyme";
import ToDoApp from "../../../components/08-useReducer/ToDoApp";

import { demoTodos } from "./fixtures/demoTodos";
import { act } from "@testing-library/react";

describe("Tests on ToDoApp", () => {

    const wrapper = shallow(<ToDoApp />);

    Storage.prototype.setItem = jest.fn();

    test("should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("should add a ToDo", () => {
        const wrapper = mount(<ToDoApp />);

        act(() => {
            wrapper.find("ToDoAdd").prop("handleAddTodo")(demoTodos[0]);
            wrapper.find("ToDoAdd").prop("handleAddTodo")(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('Todo App (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('should delete a ToDo', () => {

        wrapper.find("ToDoAdd").prop("handleAddTodo")(demoTodos[0]);
        wrapper.find("ToDoList").prop("handleDelete")(demoTodos[0].id);

        expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
    })
    
});
