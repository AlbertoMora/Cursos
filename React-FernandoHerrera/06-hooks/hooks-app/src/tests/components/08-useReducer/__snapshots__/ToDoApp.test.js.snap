// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on ToDoApp should render correctly 1`] = `
<Fragment>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ToDoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        toDos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <ToDoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
